File: src/main/java/no/fintlabs/flyt/gateway/application/archive/dispatch/file/FileDispatchService.java
package no.fintlabs.flyt.gateway.application.archive.dispatch.file;

import io.netty.handler.timeout.ReadTimeoutException;
import lombok.extern.slf4j.Slf4j;
import no.fintlabs.flyt.gateway.application.archive.dispatch.file.result.FileDispatchResult;
import no.fintlabs.flyt.gateway.application.archive.dispatch.model.instance.DokumentobjektDto;
import no.fintlabs.flyt.gateway.application.archive.dispatch.web.CreatedLocationPollTimeoutException;
import no.fintlabs.flyt.gateway.application.archive.dispatch.web.FintArchiveDispatchClient;
import no.fintlabs.flyt.gateway.application.archive.dispatch.web.flytfile.FlytFileClient;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClientResponseException;
import reactor.core.publisher.Mono;

@Service
@Slf4j
public class FileDispatchService {

    private final FintArchiveDispatchClient fintArchiveDispatchClient;
    private final FlytFileClient flytFileClient;

    public FileDispatchService(FintArchiveDispatchClient fintArchiveDispatchClient, FlytFileClient flytFileClient) {
        this.fintArchiveDispatchClient = fintArchiveDispatchClient;
        this.flytFileClient = flytFileClient;
    }

    public Mono<FileDispatchResult> dispatch(DokumentobjektDto dokumentobjektDto) {
        log.info("Dispatching file");
        return dokumentobjektDto.getFileId().map(fileId -> flytFileClient.getFile(fileId)
                        .flatMap(file -> fintArchiveDispatchClient.postFile(file)
                                .map(link -> FileDispatchResult.accepted(fileId, link))
                                .onErrorResume(
                                        WebClientResponseException.class, e -> Mono.just(
                                                FileDispatchResult.declined(fileId, e.getResponseBodyAsString())
                                        )
                                ).onErrorResume(
                                        e -> e instanceof ReadTimeoutException
                                             || e instanceof CreatedLocationPollTimeoutException,
                                        e -> {
                                            log.error("File dispatch timed out");
                                            return Mono.just(FileDispatchResult.timedOut(fileId));
                                        }
                                ).onErrorResume(e -> {
                                    log.error("File dispatch failed");
                                    return Mono.just(FileDispatchResult.failed(fileId));
                                })
                        ).onErrorResume(e -> {
                            log.error("File could not be retrieved");
                            return Mono.just(
                                    FileDispatchResult.couldNotBeRetrieved(fileId)
                            );
                        })
                ).orElse(Mono.just(FileDispatchResult.noFileId()))
                .doOnNext(result -> log.info("Dispatch result=" + result.toString()));
    }

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/dispatch/file/FileDispatchService.kt
package no.novari.flyt.archive.gateway.dispatch.file

import io.github.oshai.kotlinlogging.KotlinLogging
import no.novari.flyt.archive.gateway.dispatch.file.result.FileDispatchResult
import no.novari.flyt.archive.gateway.dispatch.isReadTimeout
import no.novari.flyt.archive.gateway.dispatch.model.instance.DokumentobjektDto
import no.novari.flyt.archive.gateway.dispatch.web.CreatedLocationPollTimeoutException
import no.novari.flyt.archive.gateway.dispatch.web.FintArchiveDispatchClient
import no.novari.flyt.archive.gateway.dispatch.web.flytfile.FlytFileClient
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClientResponseException

@Service
class FileDispatchService(
    private val fintArchiveDispatchClient: FintArchiveDispatchClient,
    private val flytFileClient: FlytFileClient,
) {
    fun dispatch(dokumentobjektDto: DokumentobjektDto): FileDispatchResult {
        log.info { "Dispatching file" }
        val fileId =
            dokumentobjektDto.fileId
                ?: return FileDispatchResult.noFileId().also { result ->
                    log.atInfo {
                        message = "Dispatch result={}"
                        arguments = arrayOf(result)
                    }
                }

        val file =
            try {
                flytFileClient.getFile(fileId)
            } catch (error: Throwable) {
                log.atError {
                    message = "File could not be retrieved"
                    cause = error
                }
                return FileDispatchResult.couldNotBeRetrieved(fileId).also {
                    log.atInfo {
                        message = "Dispatch result={}"
                        arguments = arrayOf(it)
                    }
                }
            }

        val result =
            try {
                val link = fintArchiveDispatchClient.postFile(file)
                FileDispatchResult.accepted(fileId, link)
            } catch (error: RestClientResponseException) {
                FileDispatchResult.declined(fileId, error.responseBodyAsString)
            } catch (error: CreatedLocationPollTimeoutException) {
                log.atError {
                    message = "File dispatch timed out"
                    cause = error
                }
                FileDispatchResult.timedOut(fileId)
            } catch (error: Throwable) {
                if (isReadTimeout(error)) {
                    log.atError {
                        message = "File dispatch timed out"
                        cause = error
                    }
                    FileDispatchResult.timedOut(fileId)
                } else {
                    log.atError {
                        message = "File dispatch failed"
                        cause = error
                    }
                    FileDispatchResult.failed(fileId)
                }
            }
        log.atInfo {
            message = "Dispatch result={}"
            arguments = arrayOf(result)
        }
        return result
    }

    companion object {
        private val log = KotlinLogging.logger {}
    }
}
